README.md
# healthfinance: Financial Projections and Planning for Healthcare Practices

[![Travis-CI Build Status](https://travis-ci.org/rrrlw/healthfinance.svg?branch=master)](https://travis-ci.org/rrrlw/healthfinance)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/rrrlw/healthfinance?branch=master&svg=true)](https://ci.appveyor.com/project/rrrlw/healthfinance)

[![License: GPL-3](https://img.shields.io/badge/License-GPL--3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![CRAN version](http://www.r-pkg.org/badges/version/healthfinance)](https://CRAN.R-project.org/package=healthfinance)
[![CRAN downloads](http://cranlogs.r-pkg.org/badges/grand-total/healthfinance)](https://CRAN.R-project.org/package=healthfinance)

## Overview

The goal of healthfinance is to provide a free graphical user interface for R users in healthcare management to perform basic financial modeling.

## Installation

Run the following code to install healthfinance:
```r
# install development version
devtools::install_github("rrrlw/healthfinance")

# install from CRAN (not available yet)
# install.packages("healthfinance")
```

## Functionality

Currently, the package implements a 36-month model comparing projected revenues with and without an acute economic event; this model was inspired by the COVID-19 pandemic.

## Contribute

To contribute to healthfinance, you can create issues for any bugs/suggestions on the [issues page](https://github.com/rrrlw/healthfinance/issues).
You can also fork the rrrlw/healthfinance repository and create pull requests to add features you think will be useful for useRs.

NEWS.md
# healthfinance 0.1.0

* Added Shiny UI for modeling unexpected acute economic events
* Added R function (calc_rev) for modeling unexpected acute economic events

R/utility.R
# calculate monthly revenue from initial data
# returns list containing 36 months revenue data (target & projected)
# assumes all parameters are perfect & valid - perform data validation BEFORE this
#' Calculate 3-year Revenue for Healthcare Practice
#' 
#' Allows calculation of projected revenue for upcoming 36 months along with
#' target revenue for corresponding months.
#' 
#' @param procedures df or tibble containing 3 columns (name, annual volume, annual revenue)
#' @param growth numeric vector of length 3; c(1, 10, 100) would represent
#'   expected growth of 1 percent in year 1, 10 percent in year 2
#'   (compared to year 1), and 100 percent in year 3 (compared to year 2)
#' @param comp_ratio numeric vector of length 4 containing compensation ratio
#'   (on average) of following insurances relative to Medicare: Medicare
#'   (should be 1), Medicaid, Commercial (private), and Other (self-pay, bad
#'   debt)
#' @param ins_prop numeric vector of length 4 containing proportion of
#'   patients with following types of insurance: Medicare, Medicaid,
#'   Commercial (private), and Other (self-pay, bad debt); sum of
#'   this vector should equal unity
#' @param tech_fee_mult technical fee as a multiple of procedural fee
#' @param month_prop proportion of revenue expected in each of 12 months of
#'   the year
#' @param restoration proportion of expected revenue expected in each of
#'   36 upcoming months due to acute economic event being modeled
#' @param boost_amt boost amount for up to 8 procedure sets
#' @param boost_proc list of boost procedures for each of 8 boosts above
#' @param boost_start start month (between 1 and 36, inclusive) for each of
#'   8 boosts above
#' @param boost_end end month (between 1 and 36, inclusive) for each of 8
#'   boosts above
#' @return list with 2 numeric vectors of length 36 each
#' @export
#' @examples 
#' 
#' # sample dataset of procedures
#' eg_procs <- data.frame(Name = c("Sample 1", "Sample 2", "Sample 3"),
#'                        Revenue = c(100000, 200000, 150000),
#'                        Volume = 1000, 25, 750)
#' 
#' # calculate revenue projections for next 36 months with default parameters
#' proj <- calc_rev(eg_procs)
#' 
#' # print 36-month target revenues
#' print(proj$Target)
#' 
#' # print 36-month projected revenues
#' print(proj$Projected)
calc_rev <- function(procedures,                    #- df or tibble containing 3 columns (name, annual volume, annual revenue)
                     growth = rep(0, 3),            #- default = 0% for all 3 years
                     comp_ratio = rep(1, 4),        #- Mcaid, Commercial, Other
                     ins_prop = rep(0.25, 4),       #- Mcare, Mcaid, Commercial
                     tech_fee_mult = 10,            #- encapsulates all other Mcare revenue?
                     month_prop = rep(1/12, 12),    #- default: 1/12 each month
                     restoration = rep(1, 36),      #- default: no change for 36 months
                     boost_amt = numeric(0),        #- default: no boost for all 8
                     boost_proc = list(),           #- default: nothing boosted
                     boost_start = integer(0),      #- default: start jan-2020
                     boost_end = integer(0)) {      #- default: end dec-2022
  
  # make sure parameters are valid
  # assert_that("Name" %in% colnames(procedures) &
  #               "Volume" %in% colnames(procedures) &
  #               "Revenue" %in% colnames(procedures))
  # assert_that(is.character(procedures$Name) &
  #               is.numeric(procedures$Volume) &
  #               is.numeric(procedures$Revenue))
  # assert_that(is.numeric(growth) & length(growth) == 3)
  # assert_that(is.numeric(comp_ratio()) & length(comp_ratio()) == 4)
  # assert_that(is.numeric(ins_prop) & length(ins_prop) == length(comp_ratio))
  # assert_that(is.numeric(tech_fee_mult))
  # assert_that(is.numeric(month_prop) & length(month_prop) == 12)
  # assert_that(is.numeric(restoration) & length(restoration) == 3 * length(month_prop))
  # assert_that(is.numeric(boost_amt) &
  #               is.numeric(boost_start) &
  #               is.numeric(boost_end) &
  #               is.list(boost_proc))
  # for (j in boost_proc) {
  #   assert_that(is.character(j))
  # }
  # assert_that(length(boost_amt) == length(boost_start) &
  #               length(boost_start) == length(boost_end) &
  #               length(boost_end) == length(boost_proc))
  
  # preprocessing w/ boost (remove NA)
  boost_amt[is.na(boost_amt)] <- 0
  
  procedures$Avg <- procedures$Revenue / procedures$Volume
  num_proc <- nrow(procedures)
  target_vol <- matrix(nrow = num_proc, ncol = 36)
  rownames(target_vol) <- procedures$Name
  colnames(target_vol) <- 1:36
  
  # setup monthly volumes for each year, accounting for growth
  for (i in 1:12) {
    target_vol[, i] <- procedures$Volume * month_prop[i] * (1 + growth[1] / 100)
  }
  for (i in 13:24) {
    target_vol[, i] <- target_vol[, i - 12] * (1 + growth[2] / 100)
  }
  for (i in 25:36) {
    target_vol[, i] <- target_vol[, i - 12] * (1 + growth[3] / 100)
  }
  
  # account for insurance + tech fee reimbursement multiplier
  reimb_mult <- sum(comp_ratio * ins_prop) * (1 + tech_fee_mult)
  target_vol <- target_vol * reimb_mult
  
  # make modifications for projected volume based on boost + restoration
  project_vol <- target_vol
  
  # account for restoration
  for (i in 1:36) {
    project_vol[, i] <- project_vol[, i] * restoration[i]
  }
  
  # add boosts
  if (length(boost_proc) > 0) {
    for (i in 1:length(boost_proc)) {
      curr_start <- boost_start[i]
      curr_end <- boost_end[i]
      curr_amt <- boost_amt[i]
      for (j in boost_proc[[i]]) {
        if (!(j %in% rownames(project_vol))) {
          print(paste(j, "not valid procedure, skipping"))
          next
        }
        project_vol[j, curr_start:curr_end] <- project_vol[j, curr_start:curr_end] * (1 + curr_amt)
      }
    }
  }
  
  # convert volumes to revenues
  for (i in 1:nrow(target_vol)) {
    target_vol[i, ] <- target_vol[i, ] * procedures$Avg[i]
    project_vol[i, ]<- project_vol[i, ] * procedures$Avg[i]
  }
  
  # sum up revenues by month
  target_rev <- numeric(ncol(target_vol))
  project_rev<- numeric(ncol(project_vol))
  for (i in 1:ncol(target_vol)) {
    target_rev[i] <- sum(target_vol[, i])
    project_rev[i]<- sum(project_vol[, i])
  }
  
  list("Target" = target_rev,
       "Projected" = project_rev)
}

inst/hfin/global.R
#####SETUP#####
# library("tidyverse")

# utility functions on which workhorses rely

#####VALIDATION FUNcTIONS#####
# make sure item info is valid for addition
assert_rowadd <- function(name, vol, rev, names) {
  
}

# make sure row number is valid
assert_rownum <- function(rownum) {
  
}

# make sure input file is valid
assert_inputfile <- function(file_path) {
  
}

# make sure all parameters are valid
assert_params <- function(procedures, growth, comp_ratio, ins_prop,
                          tech_fee_mult, month_prop, restoration, boost_amt,
                          boost_proc, boost_start, boost_end) {
  
}

#####UTILITY FUNCTIONS#####
# assume params already assert/validated
calc_reimburse_mult <- function(comp_ratio, #compensation relative to Mcare
                                ins_prop,   #pt payor mix
                                tech_fee_ratio) {
  sum(comp_ratio() * ins_prop()) * (1 + tech_fee_ratio())
}

# constants for shiny app
CURRENCY_DIGITS <- 13
MODEL_MONTHS <- 36

# function to customize ggplot2 axis lables
custom_dollar <- function(x) {
  scales::dollar(x,
                 scale = 1e-6,
                 largest_with_cents = 0,
                 suffix = "M")
}
text_to_currency <- function(num) {
  # check parameter validity
  if (!is.numeric(num)) stop(paste("Invalid parameter passed to text_to_currency:",
                                   num))
  
  # preserve options
  temp_prev <- options()$scipen
  options(scipen = CURRENCY_DIGITS)
  
  # convert to text
  ans <- as.character(round(num, 2))
  
  # revert options and return answer
  options(scipen = temp_prev)
  return(ans)
}
str_to_vec_year <- function(txt) {
  ans <- as.numeric(
    strsplit(x = txt,
             split = ",",
             fixed = TRUE)[[1]]
  )
  # confirm there's exactly 12 non-NA, numeric elements
  if (length(ans) == 12 &
      sum(is.na(ans)) == 0) {
    return(ans)
  } else {
    stop(paste("Problem with passed parameter:",
               txt))
  }
}

# export model data to CSV file
export_model <- function(df, path) {
  if (file.exists(path)) {
    readr::write_csv(df, path, append = TRUE)
  } else {
    readr::write_csv(df, path)
  }
}

inst/hfin/ui.R
#####VIEW#####
fluidPage(
  tagList(
    tags$style("# params { display:none; }"),
    tabsetPanel(id = "params",
      # input for the model (file or manual)
      tabPanel("Import",
        fluidRow(
          column(4, fileInput("input", label = "")),
          column(2, actionButton("readfile", "Import Data")),
          column(2, htmlOutput("rownumdeltext")),
          column(2, numericInput("rownumdel", "", value = 1,
                                 min = 1)),
          column(2, actionButton("delrow", "Delete Item"))
        ),
        fluidRow(
          column(2, htmlOutput("rowaddtext")),
          column(2, textInput("addname", "Item Name")),
          column(2, numericInput("addvol", "Item Volume",
                                 value = 0, min = 0)),
          column(2, numericInput("addrev", "Item Revenue",
                                 value = 0, min = 0)),
          column(2, actionButton("addrow", "Add Item"))
        ),
        tableOutput("currinput")
      ),
      # financial model (not model from model-view-adapter)
      tabPanel("Model",
        fluidRow(
          column(3, htmlOutput("targetrev")),
          column(3, htmlOutput("projrev")),
          column(6, plotOutput("project12", height = "225px"))
        ),
        fluidRow(
          column(6, plotOutput("project24", height = "225px")),
          column(6, plotOutput("project36", height = "225px"))
        ),
        
        #####UI: GROWTH#####
        fluidRow(
          # column(12, verbatimTextOutput("growthheader")),
          column(3, actionButton("calc", "Calculate")),
          column(3, sliderInput("growth20", "Growth 2019-20", 
                                min = -10, max = 10, value = 0)),
          column(3, sliderInput("growth21", "Growth 2020-21",
                                min = -10, max = 10, value = 0)),
          column(3, sliderInput("growth22", "Growth 2021-22",
                                min = -10, max = 10, value = 0))
        ),
        
        #####UI: INSURANCE#####
        fluidRow(column(12, htmlOutput("instxt"))),
        fluidRow(
          column(3, "Types:"),
          column(2, "Medicare"),
          column(2, "Medicaid"),
          column(2, "Commercial"),
          column(2, "Other")
        ),
        fluidRow(
          column(3, "Compensation Ratio:"),
          column(2, "1"),  # ratio of medicaid compensation to itself is always 1
          column(2, numericInput("mcaidcomp", NULL, value = 0.63, min = 0)),
          column(2, numericInput("commcomp", NULL, value = 2.25, min = 0)),
          column(2, numericInput("noinscomp", NULL, value = 0, min = 0))
        ),
        fluidRow(
          column(3, "Proportion:"),
          column(2, numericInput("mcareprop", NULL, value = 0.47, min = 0, max = 1)),
          column(2, numericInput("mcaidprop", NULL, value = 0.14, min = 0, max = 1)),
          column(2, numericInput("commprop", NULL, value = 0.37, min = 0, max = 1)),
          column(2, textOutput("noinsprop"))
        ),
        # technical fee slider
        fluidRow(
          column(12, sliderInput("techratio", "Facility Fee Multiplier",
                        min = 0, max = 25, value = 10, width = "100%", step = 0.5))
        ),
        
        #####UI: ANNUAL BREAKDOWN#####
        fluidRow(
          column(12, htmlOutput("annualtxt")),
          column(2, numericInput("annualm1", "January",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm2", "February",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm3", "March",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm4", "April",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm5", "May",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm6", "June",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm7", "July",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm8", "August",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm9", "September",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm10", "October",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, numericInput("annualm11", "November",
                                 value = round(1/12, 5), min = 0, max = 1)),
          column(2, htmlOutput("annualm12"))
        ),
        
        #####UI: ACTIVITY RESTORATION#####
        fluidRow(
          column(12, htmlOutput("restoretxt")),
          column(4, textInput("restore12", "Restoration: 1-12 months",
                              value = paste(rep("1", 12), collapse = ","),
                              width = "100%")),
          column(4, textInput("restore24", "Restoration: 13-24 months",
                              value = paste(rep("1", 12), collapse = ","),
                              width = "100%")),
          column(4, textInput("restore36", "Restoration: 25-36 months",
                              value = paste(rep("1", 12), collapse = ","),
                              width = "100%"))
        ),
        
        #####UI: BOOST PROCEDURES#####
        fluidRow(
          # boost 1-4 amounts
          column(3, numericInput("boost1", "Boost #1 Amount", value = 0)),
          column(3, numericInput("boost2", "Boost #2 Amount", value = 0)),
          column(3, numericInput("boost3", "Boost #3 Amount", value = 0)),
          column(3, numericInput("boost4", "Boost #4 Amount", value = 0)),
          
          # boost 1-4 procedures
          column(3, selectInput("boostproc1", "Boost #1 Procedures",
                                choices = c("Sample Procedure 1",
                                            "Sample Procedure 2",
                                            "Sample Procedure 3"), multiple = TRUE)),
          column(3, selectInput("boostproc2", "Boost #2 Procedures",
                                choices = c("Sample Procedure 1",
                                            "Sample Procedure 2",
                                            "Sample Procedure 3"), multiple = TRUE)),
          column(3, selectInput("boostproc3", "Boost #3 Procedures",
                                choices = c("Sample Procedure 1",
                                            "Sample Procedure 2",
                                            "Sample Procedure 3"), multiple = TRUE)),
          column(3, selectInput("boostproc4", "Boost #4 Procedures",
                                choices = c("Sample Procedure 1",
                                            "Sample Procedure 2",
                                            "Sample Procedure 3"), multiple = TRUE)),
          
          # boost 1-4 durations
          column(3, dateRangeInput("boostdur1", "Boost #1 Duration",
                                   start = "2020-01-01", end = "2022-12-31",
                                   min = "2020-01-01", max = "2022-12-31",
                                   format = "M-yyyy")),
          column(3, dateRangeInput("boostdur2", "Boost #2 Duration",
                                   start = "2020-01-01", end = "2022-12-31",
                                   min = "2020-01-01", max = "2022-12-31",
                                   format = "M-yyyy")),
          column(3, dateRangeInput("boostdur3", "Boost #3 Duration",
                                   start = "2020-01-01", end = "2022-12-31",
                                   min = "2020-01-01", max = "2022-12-31",
                                   format = "M-yyyy")),
          column(3, dateRangeInput("boostdur4", "Boost #4 Duration",
                                   start = "2020-01-01", end = "2022-12-31",
                                   min = "2020-01-01", max = "2022-12-31",
                                   format = "M-yyyy")),
          
          # boost 5-8 amounts
          column(3, numericInput("boost5", "Boost #5 Amount", value = 0)),
          column(3, numericInput("boost6", "Boost #6 Amount", value = 0)),
          column(3, numericInput("boost7", "Boost #7 Amount", value = 0)),
          column(3, numericInput("boost8", "Boost #8 Amount", value = 0)),
          
          # boost 5-8 procedures
          column(3, selectInput("boostproc5", "Boost #5 Procedures",
                                choices = c("Sample Procedure 1",
                                            "Sample Procedure 2",
                                            "Sample Procedure 3"), multiple = TRUE)),
          column(3, selectInput("boostproc6", "Boost #6 Procedures",
                                choices = c("Sample Procedure 1",
                                            "Sample Procedure 2",
                                            "Sample Procedure 3"), multiple = TRUE)),
          column(3, selectInput("boostproc7", "Boost #7 Procedures",
                                choices = c("Sample Procedure 1",
                                            "Sample Procedure 2",
                                            "Sample Procedure 3"), multiple = TRUE)),
          column(3, selectInput("boostproc8", "Boost #8 Procedures",
                                choices = c("Sample Procedure 1",
                                            "Sample Procedure 2",
                                            "Sample Procedure 3"), multiple = TRUE)),
          
          # boost 5-8 durations
          column(3, dateRangeInput("boostdur5", "Boost #5 Duration",
                                   start = "2020-01-01", end = "2022-12-31",
                                   min = "2020-01-01", max = "2022-12-31",
                                   format = "M-yyyy")),
          column(3, dateRangeInput("boostdur6", "Boost #6 Duration",
                                   start = "2020-01-01", end = "2022-12-31",
                                   min = "2020-01-01", max = "2022-12-31",
                                   format = "M-yyyy")),
          column(3, dateRangeInput("boostdur7", "Boost #7 Duration",
                                   start = "2020-01-01", end = "2022-12-31",
                                   min = "2020-01-01", max = "2022-12-31",
                                   format = "M-yyyy")),
          column(3, dateRangeInput("boostdur8", "Boost #8 Duration",
                                   start = "2020-01-01", end = "2022-12-31",
                                   min = "2020-01-01", max = "2022-12-31",
                                   format = "M-yyyy"))
        )
      ),
      tabPanel("Export",
        fluidRow(
          column(4, textInput("modelname", "Model Name")),
          column(4, textInput("filename", "File Name")),
          column(1, actionButton("export", "Export Parameters"))
        )
      )
    )
  )
)
